quicktype-core 24.0.0: 'lang' is now a strict literal union — README-style samples using 'string' no longer typecheck; document LanguageName

## Issue Type

DX regression / documentation gap in the `quicktype-core` API

## Context (Environment, Version, Language)

CLI, npm, or app.quicktype.io: npm package `quicktype-core`
Version: 24.0.0
TypeScript: 5.8.3

## Description

In 24.0.0 the `lang` option of `quicktype()` (and the parameter of `jsonInputForTargetLanguage()`) is typed as a strict literal union (`LanguageName | TargetLanguage`) instead of `string`.

The widely-circulated usage sample (README / blog posts / Stack Overflow answers) wraps quicktype in a helper taking `targetLanguage: string`:

```ts
async function quicktypeJSON(targetLanguage: string, typeName: string, jsonString: string) {
    const jsonInput = jsonInputForTargetLanguage(targetLanguage);
    // ...
    return quicktype({ inputData, lang: targetLanguage });
}
```

This no longer typechecks:

```
error TS2345: Argument of type 'string' is not assignable to parameter of type '"ruby" | TargetLanguage<LanguageConfig> | "cjson" | "cJSON" | "c++" | ...'.
error TS2322: Type 'string' is not assignable to type '"ruby" | TargetLanguage<LanguageConfig> | ... | undefined'.
```

The fix is to use the exported `LanguageName` type:

```ts
import { type LanguageName } from "quicktype-core";
async function quicktypeJSON(targetLanguage: LanguageName, ...) { ... }
```

The stricter typing is arguably an improvement (typos in language names are caught at compile time), but it's a breaking change for anyone porting existing sample code, and `LanguageName` doesn't appear in the docs/README samples.

## Expected Behaviour / Output

Either:
- documented migration guidance: update the README/sample code to use `LanguageName`, and mention the break in the 24.0.0 release notes, and/or
- an escape hatch for dynamic language names (e.g. keep `isLanguageName()` prominent in the docs so `string` values can be narrowed before the call).

## Current Behaviour / Output

Code following the published examples fails to compile against 24.0.0 with TS2345/TS2322, and the resolution (`LanguageName`) is only discoverable by reading `dist/language/types.d.ts`.

## Steps to Reproduce

```bash
npm install quicktype-core@24.0.0 typescript@5.8
# compile the README-style helper above with tsc --strict
```

Found while testing the 24.0.0 release on Node 20/22/24.

--- a/README.md
+++ b/README.md
@@ -189,6 +189,25 @@ main();
 
 The argument to `quicktype` is a complex object with many optional properties. [Explore its definition](https://github.com/quicktype/quicktype/blob/master/packages/quicktype-core/src/Run.ts#L637) to understand what options are allowed.
 
+#### TypeScript: the `lang` option is a `LanguageName`, not a `string`
+
+In TypeScript, the `lang` option of `quicktype` and the argument of `jsonInputForTargetLanguage` are typed as `LanguageName`, a union of all supported language names — passing a plain `string` is a compile-time error. If the language name is known statically, just pass the literal (`lang: "swift"`). For wrappers that take the language name at runtime — from CLI arguments or configuration — validate it with the `isLanguageName` type guard instead of casting:
+
+```typescript
+import { type LanguageName, isLanguageName, quicktype } from "quicktype-core";
+
+async function quicktypeJSON(targetLanguage: LanguageName, typeName: string, jsonString: string) {
+    // same body as above
+}
+
+const lang = process.argv[2]; // untrusted string
+if (!isLanguageName(lang)) {
+    throw new Error(`Unknown language: ${lang}`);
+}
+// lang now has type LanguageName
+const { lines } = await quicktypeJSON(lang, "Person", jsonString);
+```
+
 #### Bundling `quicktype-core` as ESM
 
 `quicktype-core` ships both CommonJS and ESM builds. If you bundle it into an ESM output for Node with esbuild (`--format=esm --platform=node`), the bundle may throw at runtime:

--- a/packages/quicktype-core/src/language/All.ts
+++ b/packages/quicktype-core/src/language/All.ts
@@ -67,6 +67,14 @@ export const all = [
 
 all satisfies readonly TargetLanguage[];
 
+/**
+ * Returns the `TargetLanguage` registered under `name`.
+ *
+ * `name` must be one of the `LanguageName` literals, so plain strings from
+ * CLI arguments or configuration must first be narrowed with
+ * {@link isLanguageName}.  Throws if no language in `targetLanguages` is
+ * named `name`, which can only happen when passing a custom language list.
+ */
 export function languageNamed<Name extends LanguageName>(
     name: Name,
     targetLanguages: readonly TargetLanguage[] = all,
@@ -81,6 +89,21 @@ export function languageNamed<Name extends LanguageName>(
     return foundLanguage as LanguageNameMap[Name];
 }
 
+/**
+ * Type guard that narrows a runtime `string` to `LanguageName`.
+ *
+ * Use this to validate untrusted input — CLI arguments, configuration
+ * files, HTTP parameters — before passing it to `quicktype` or
+ * `jsonInputForTargetLanguage`, instead of casting with `as LanguageName`.
+ *
+ * @example
+ * const lang = process.argv[2];
+ * if (!isLanguageName(lang)) {
+ *     throw new Error(`Unknown language: ${lang}`);
+ * }
+ * // lang now has type LanguageName
+ * await quicktype({ inputData, lang });
+ */
 export function isLanguageName(maybeName: string): maybeName is LanguageName {
     if (
         all.some((lang) =>

--- a/packages/quicktype-core/src/language/types.ts
+++ b/packages/quicktype-core/src/language/types.ts
@@ -4,9 +4,18 @@ import type { all } from "./All.js";
 
 type AllLanguages = (typeof all)[number];
 
+/** The display name of a built-in target language, e.g. `"C++"` or `"TypeScript"`. */
 export type LanguageDisplayName<
     Language extends TargetLanguage = AllLanguages,
 > = Language["displayName"];
+/**
+ * The union of all name aliases of the built-in target languages, e.g.
+ * `"c++"`, `"ts"`, `"typescript"`.  This is what `quicktype`'s `lang`
+ * option and `jsonInputForTargetLanguage` accept instead of `string`.
+ *
+ * To use a runtime `string` where a `LanguageName` is expected, narrow
+ * it with the `isLanguageName` type guard rather than casting.
+ */
 export type LanguageName<Language extends TargetLanguage = AllLanguages> =
     Language["names"][number];
 

